packages/winpanda/extra/src/winpanda/svcm/nssm.py

```diff
@@ -0,0 +1,402 @@
+"""Winpanda: Windows service management: NSSM-based manager definition.
+
+Ref:
+  [1] NSSM - the Non-Sucking Service Manager
+      https://nssm.cc/description
+  [2] nssm/README.txt
+      https://git.nssm.cc/nssm/nssm/src/master/README.txt
+"""
+import configparser
+import enum
+import os
+from pathlib import Path
+import re
+import subprocess
+
+import jinja2 as jj2
+
+from . import base
+from . import exceptions as svcm_exc
+import constants as const
+from core import logger
+
+
+LOG = logger.get_logger(__name__)
+
+
+class NSSMParameter(enum.Enum):
+    """NSSM parameter set."""
+    DESCRIPTION = 'Description'
+    DISPLAYNAME = 'DisplayName'
+    NAME = 'Name'
+    APPLICATION = 'Application'
+    APPDIRECTORY = 'AppDirectory'
+    APPPARAMETERS = 'AppParameters'
+    START = 'Start'
+    DEPENDONSERVICE = 'DependOnService'
+    APPSTDOUT = 'AppStdout'
+    APPSTDERR = 'AppStderr'
+    APPENVIRONMENTEXTRA = 'AppEnvironmentExtra'
+
+    @classmethod
+    def values(cls):
+        return [m.value for m in cls.__members__.values()]
+
+    @classmethod
+    def values_required(cls):
+        # Names of required parameters. !!!Please keep the sequence!!!
+        names_required = ('DISPLAYNAME', 'APPLICATION')
+        return [
+            m.value for n, m in cls.__members__.items() if n in names_required
+        ]
+
+
+class NSSMCommand(enum.Enum):
+    """NSSM command set."""
+    INSTALL = 'install'
+    REMOVE = 'remove'
+    START = 'start'
+    STOP = 'stop'
+    RESTART = 'restart'
+    STATUS = 'status'
+    SET = 'set'
+
+    @classmethod
+    def values(cls):
+        return [m.value for m in cls.__members__.values()]
+
+    @classmethod
+    def values_primitive(cls):
+        # Names of primitive commands.
+        names_primitive = ('START', 'STOP', 'RESTART', 'STATUS')
+        return [
+            m.value for n, m in cls.__members__.items() if n in names_primitive
+        ]
+
+
+class NSSMConfSection(enum.Enum):
+    """NSSM configuration file sections."""
+    SERVICE = 'service'
+
+
+@base.svcm_type('nssm')
+class WinSvcManagerNSSM(base.WindowsServiceManager):
+    """NSSM-based Windows service manager."""
+    __exec_fname__ = 'nssm.exe'  # Executable file name.
```

Comment:
What is the intent with the underscores here? General Python behaviour is:
- a single leading underscore indicates a private attribute, but it is not enforced by the interpreted
- a double leading underscore indicates a private attribute, somewhat enforced by being obscured through name mangling
- double leading and trailing underscores are for Python language attributes, not for general use (and are not private in any way).

Reply:
That's correct. Will fix it.

Reply:
Fixed.